### PR TITLE
[Nav] Zed gimbal addition

### PR DESCRIPTION
**Pull request by ankithu**
#388 , #389

This commit adds a Gimbal class to allow control of the Zed Gimbal for improved search efficiency. It also adds the use of that gimbal class through added states to both the Gate Traversal and Search state machines. The basic functionality is that instead of turning 360 degrees when we get to a point we want to search. We turn the gimbal to positive 150 degrees then to -150 degrees then back to 0. This number is editable through the config. Additionally, code was added to receive and send gimbal LCM messages. Additionally, as this is a new feature, a boolean was added to the config that can be turned on or off to enable or disable Zed Gimbal functionality.

Changes were tested in the simulator. The search state machine changes were tested with various locations of waypoints and AR tags, in all instances that I tested, the rover was able to successfully find the AR tags. The gate traversal algorithm also worked with various orientations of the gates and distances from the point (enough to enter the diamond search pattern) and the rover made it through even with the gate pointing in a direction that would be opposite from ideal. Additionally, the enable and disable gimbal switch was tested and works in the sim.

### CODE REVIEW

**Comment by ankithu** [comment]
Hi, just a quick question, what would be the best way to resolve the merge conflicts? Manually or should I merge in the auton integration branch with this branch or maybe rebase zed_gimbal_addition on top of the auton_integration branch?

**Comment by aswarner** [comment]
I think you should rebase auton_integration under your changes, then you will probably need to manually resolve remaining conflicts in your editor. I am surprised there were issues since you pulled from auton_integration to start working with Zack's code. But that's okay. update your PR after fixing your conflicts

**Comment by aswarner** [comment]
also, remove amverni from the review since he is retired. and you can keep James on, but since he worked with you he shouldn't be a primary reviewer. you should have 
1. a member familiar with the specific code in question (since this is nav code, you should add me for this)
2. a member familiar with MRover coding standards and review process (david smith @DavidSmith166 or raytitan) 
3. a new member (I think we suggested Rajin, but it can be anyone who will be responsive)

**Comment by ankithu** [comment]
Ok merge conflicts are fixed and I adjusted the reviewers (minus Rajin, waiting on a username)

**Comment by aswarner** [comment]
@ankithu  nice job fixing the merge conflicts! did you test after making the changes?

**Comment by ankithu** [comment]
Ok just tested and everything seems to work after a silly mistake was fixed

**Review by aswarner — CHANGES_REQUESTED** [review]
We need to make two major changes: avoiding using "target yaw/angle" in gimbal functions and variables, and refactoring gateSpinWait functions and searchSpinWait/searchGimbalWait into just gateWait and searchWait.

Buggy simulator code is still in autonomous integration branch - I will revert, then we need to pull the modified files from this PR

File: jetson/nav/gate_search/gateStateMachine.cpp [hunk 1]
```diff
@@ -112,13 +117,80 @@ NavState GateStateMachine::executeGateSpin()
...
+    //if target aquired, go to it
+    if( mPhoebe->roverStatus().rightTarget().distance >= 0 ||
+        ( mPhoebe->roverStatus().leftTarget().distance >= 0 && mPhoebe->roverStatus().leftTarget().id != lastKnownRightPost.id ) )
+    {
+        updatePost2Info();
+        calcCenterPoint();
+        return NavState::GateTurnToCentPoint;
+    }
+
+    //set the target yaw to wherever the next stop on the gimbals path is
```
Comment (aswarner) [inline jetson/nav/gate_search/gateStateMachine.cpp:145]: change comments in this function to say desired/phase etc, not target

File: jetson/nav/gimbal.hpp [hunk 2]
```diff
@@ -0,0 +1,33 @@
...
+
+class Gimbal{
+    private:
+        double cur_yaw;
+        double desired_gimbal_yaw;
+        double TOLERANCE;
+        rover_msgs::ZedGimbalPosition signal;
+    public:
+        Gimbal(double tolerance_in );
+        //sets the target yaw of the gimbal
```
Comment (aswarner) [inline jetson/nav/gimbal.hpp:20]: remove "target"

File: jetson/nav/gimbal.hpp [hunk 3]
```diff
@@ -0,0 +1,33 @@
...
+        bool setDesiredGimbalYaw( double target );
```
Comment (aswarner) [inline jetson/nav/gimbal.hpp:21]: takes input "target"  - remove

**Review by aswarner — CHANGES_REQUESTED** [review]

File: simulators/nav/src/store/modules/simulatorState.ts [hunk 4]
```diff
@@ -18,6 +17,7 @@ import {
 /**************************************************************************************************
  * Constants
  **************************************************************************************************/
+// const MAX_RECENT_PATH_LEN = 100;
```
Comment (John-Trager) [inline simulators/nav/src/store/modules/simulatorState.ts:20]: does this need to stay in?

Comment (ankithu) [reply simulators/nav/src/store/modules/simulatorState.ts:20]: discussed offline

**Comment by aswarner** [comment]
can you remove the extra waits in between reaching the first 150 degrees and after reaching the -150? e.g. remove redundant waits at the same angle

**Review by aswarner — APPROVED** [review]
looks ready to me! I made two small changes. other reviewers, you will need to modify the json file to make useGimbal true before testing. I made the default false (use spin for search).

**Review by John-Trager — APPROVED** [review]
Reviewed code and ran in sim. No problems were found looks good.

**Review by raytitan — APPROVED** [review]
Let's get this tested!

**Pull request merged**